example/z-docker/Dockerfile

```diff
@@ -1,4 +1,40 @@
-FROM ubuntu:latest
-RUN apt update && apt install -y openssl libev4
-COPY app.exe .
-ENTRYPOINT /app.exe
+FROM ocaml/opam:alpine as build
+
+# Install system dependencies
+# Some of these are only needed for database and JavaScript stuff, but this is 
+# our build context, it won't impact our runtime environment.
+RUN sudo apk add --update \
+    make libffi-dev linux-headers m4 ncurses-dev perl pkgconf postgresql-dev \
+    musl-dev git pkgconfig gmp-dev libc-dev openblas-dev zlib-dev \
+    npm nodejs libev-dev 
+
+# Install Esy-specific dependencies
+RUN sudo apk add \
+    perl-utils
+
+RUN eval $(opam env) && opam repository set-url default https://opam.ocaml.org && opam update
+
+WORKDIR /build
+
+RUN npm install esy
+
+# Install dependencies
+# We only copy the opam file to install dependencies only when they change, and 
+# not every time another file is modified.
+ADD esy.json esy.json
+ADD esy.lock esy.lock
+RUN node_modules/.bin/esy fetch
+RUN node_modules/.bin/esy build-dependencies
+
+# Build project
+ADD . .
+
+RUN node_modules/.bin/esy build
+
+FROM alpine:3.12 as run
+
+RUN apk add --update gmp libpq libev
```

Comment:
With esy we can get these from the esy store to make sure they work even if we would want to use `scratch`

Reply:
I'm not familiar with Esy, what would the commands to install them look like?

Reply:
esy installs them automatically. It essentially vendors a bunch of deps that would otherwise be system deps.

Compare with the instructions for this example before the PR. They give the set of necessary system deps. Everything else is installed by esy, including libev, gmp, and more.

Reply:
I think this comment is targetting the build stage then, right? The runtime dependencies are needed even with Esy (I tested and have symbol not found errors when not installing `libev`)

Reply:
> I think this comment is targetting the build stage then, right?

This comment was about targetting the run-time stage, not the build stage. If you looked in the original, "run-time" `Dockerfile` for this example, as I suggested, you would see the necessary system deps I was referring to:

https://github.com/aantron/dream/blob/67842e7b92f5b76058abe877cea63b89b31704af/example/z-docker/Dockerfile#L2

> The runtime dependencies are needed even with Esy (I tested and have symbol not found errors when not installing `libev`)

So, it's not surprising that testing only removing specifically libev was causing a failure &mdash; it is required even with esy, as was already known. You would have had to test the other dependencies. Your original code that was at issue in this conversation included several other libraries, like gmp and libpq, which is what prompted the discussion. However, it appears you have since slimmed them down, so this is not an issue anymore. You implicitly "tested" what me and @ulrikstrid were getting at by successfully removing all of them.

Reply:
Ok ok, thanks for the defailed explanation 🙂

Reply:
I have a example somewhere on how to copy  from the esy store and reuse them in the target image